Snakefile

```diff
@@ -1143,32 +1150,68 @@ rule encode_sentencepiece_ids:
             " --max-length {MAX_SPM_ENCODE_IDS}"
             " --file-prefix {output_prefix}"
         )
 
 
+# Split the encoded SentencePiece id's into training, validation, and testing sets.
+rule split_one_sentencepiece_id_file:
+    input:
+        "data/sentencepiece/{lang}.spm_txt_ids/length_{n}.txt"
+    output:
+        temp("data/sentencepiece/{lang}.spm_txt_ids_training/length_{n}.txt"),
+        temp("data/sentencepiece/{lang}.spm_txt_ids_validation/length_{n}.txt"),
+        temp("data/sentencepiece/{lang}.spm_txt_ids_testing/length_{n}.txt")
+    run:
+        training_file, validation_file, testing_file = output
+        frac_train = FRACTION_OF_ENCODED_SPM_DATA_FOR_TRAINING
+        frac_valid = FRACTION_OF_REMAINING_SPM_DATA_FOR_VALIDATION
+        # touch the outputs so that they are created even if empty
+        shell(
+            "cat {input} | "
+            "python scripts/split_sentencepiece_id_file.py"
+            "  --fraction {frac_train}"
+            "  --seed 1"
+            "  --output-file {training_file} | "
+            "python scripts/split_sentencepiece_id_file.py"
+            "  --fraction {frac_valid}"
+            "  --seed 2"
+            "  --output-file {validation_file} > "
+            "{testing_file} ; "
+            "touch {training_file} ; "
+            "touch {validation_file} ; "
+            "touch {testing_file}"
+        )
+
+
 # Convert the outputs of encode_sentencepiece_ids into .npy files.
 rule numberize_one_sentencepiece_id_file:
     input:
-        "data/sentencepiece/{lang}.spm_txt_ids/length_{n}.txt"
+        "data/sentencepiece/{lang}.spm_txt_ids_{task}/length_{n}.txt"
     output:
-        "data/sentencepiece/{lang}.spm_ids/length_{n}.npy"
+        "data/sentencepiece/{lang}.spm_ids_{task}/length_{n}.npy"
     shell:
         "scripts/sentencepiece_id_converter.py {input} {output}"
 
 
 rule numberize_sentencepiece_ids:
     input:
-        expand("data/sentencepiece/{{lang}}.spm_ids/length_{n}.npy",
-               n=range(MAX_SPM_ENCODE_IDS + 1))
+        expand("data/sentencepiece/{{lang}}.spm_ids_{task}/length_{n}.npy",
+               n=range(MAX_SPM_ENCODE_IDS + 1),
+               task=["training", "validation", "testing"])
     output:
-        "data/sentencepiece/{lang}.spm_ids/empty_build_target"
+        touch("data/sentencepiece/{lang}.spm_ids_training/empty_build_target"),
+        touch("data/sentencepiece/{lang}.spm_ids_validation/empty_build_target"),
+        touch("data/sentencepiece/{lang}.spm_ids_testing/empty_build_target")
     shell:
-        "rmdir data/sentencepiece/{wildcards.lang}.spm_txt_ids; touch {output}"
+        "rmdir data/sentencepiece/{wildcards.lang}.spm_txt_ids_training ; "
+        "rmdir data/sentencepiece/{wildcards.lang}.spm_txt_ids_validation ; "
+        "rmdir data/sentencepiece/{wildcards.lang}.spm_txt_ids_testing ; "
+        "rmdir data/sentencepiece/{wildcards.lang}.spm_txt_ids"
```

Comment:
I know we had something like this rule before, but what does it accomplish? I don't see it as essential to clean up empty directories, particularly not when it needs to create other empty files to show that it's done.

It looks to me like this should be a top-level rule, expanded on both `{n}` and `{lang}`, with a more memorable name, and without the "output" and "shell" sections.

Reply:
I can make this change, but it is far from clear to me what languages should be used to do the expansion.  There are several language lists defined in the Snakefile, but none of them seem to correspond to the data available in data/monolingual (which right now just seems to have data for English).  Should I define a new list just containing English (for now at least), or is there some better way?

Reply:
OK, this rule has been replaced by a top-level rule expanded on the language (using a new list of SentencePiece languages, currently just English), sequence length, and task (training, testing, validation).